ITensorGaussianMPS/src/gmps.jl

```diff
@@ -188,6 +294,193 @@ function givens_rotations(v::AbstractVector{ElT}) where {ElT}
+  gsac = Circuit(copy(gsaa.rotations))
+  gsac = shift!(gsac, +1)
+  gsac = conj!(gsac)
+  gsa = interleave(gsac, gsaa)
+  v = gsa * v
+  LinearAlgebra.lmul!(gsa, gs)
+
+  ##Boguliobov rotation for remaining Bell pair
+  g1, r = givens(v, 2, 3)
+  g2 = Givens(1, 4, g1.c, g1.s')
+  v = g1 * v
+  v = g2 * v #should have no effect
+  LinearAlgebra.lmul!(g2, gs)
+  LinearAlgebra.lmul!(g1, gs)
+  return gs, r
+end
+
+function check_pairing_correlations(Λ0::AbstractMatrix{ElT}) where {ElT<:Number}
+  paired = false
+  Λblocked = reverse_interleave(Λ0)
+  N = div(size(Λblocked, 1), 2)
+  if all(abs.(Λblocked[1:N, (N + 1):end]) .<= eps(Float64))
+    return paired, Λblocked[(N + 1):end, (N + 1):end]
+  else
+    paired = true
+    return paired, Λ0
+  end
+end
+
+function check_pairing_correlations(Λ0::ConservingNf)
+  return Λ0
+end
+
+function check_pairing_correlations(Λ0::Pairing)
+  is_paired, Λ = check_pairing_correlations(Λ0.data)
+  if !is_paired
+    return ConservingNf(Λ)
+  else
+    return Pairing(Λ)
+  end
+end
+
+function get_subblock(
+  _Λ::Pairing,
+  startind::Int;
+  eigval_cutoff::Float64=1e-8,
+  minblocksize::Int=1,
+  maxblocksize::Int=div(size(_Λ.data, 1), 2),
+)
+  blocksize = 0
+  n = 0.0
+  n1 = 0.0
+  n2 = 0.0
+  err = 0.0
+  p = Int[]
+  nB = 0.0
+  uB = 0.0
+  ΛB = 0.0
+  i = startind
+  Λ = _Λ.data
+  N = size(Λ, 1)
+  for blocksize in minblocksize:maxblocksize
+    j = min(2 * i + 2 * blocksize, N)
+    ΛB = @view Λ[(2 * i - 1):j, (2 * i - 1):j]
+    nB, uB = eigen(Hermitian(ΛB))
+    p = sortperm(nB)
+    n1 = nB[first(p)]
+    n2 = nB[last(p)]
+    err = min(abs(n1), abs(n2))
+    n = n1
+    err ≤ eigval_cutoff && break
+  end
+  v = @view uB[:, p[1]]
+  return Pairing(v), Pairing(nB), err
+end
+
+function get_subblock(
+  _Λ::ConservingNf,
+  startind::Int;
+  eigval_cutoff::Float64=1e-8,
+  minblocksize::Int=1,
+  maxblocksize::Int=size(_Λ.data, 1),
+)
+  blocksize = 0
+  n = 0.0
+  err = 0.0
+  p = Int[]
+  nB = 0.0
+  uB = 0.0
+  ΛB = 0.0
+  i = startind
+  Λ = _Λ.data
+  N = size(Λ, 1)
+  for blocksize in minblocksize:maxblocksize
+    j = min(i + blocksize, N)
+    ΛB = @view Λ[i:j, i:j]
+    nB, uB = eigen(Hermitian(ΛB))
+    p = sortperm(nB; by=entropy)
+    n = nB[p[1]]
+    err = min(n, 1 - n)
+    err ≤ eigval_cutoff && break
+  end
+  v = @view uB[:, p[1]]
+  return ConservingNf(v), ConservingNf(nB), err
+end
+
+function process_populations!(
+  _ns::ConservingNf, _nB::ConservingNf, _v::ConservingNf, i::Int
+)
+  p = Int[]
+  ns = _ns.data
+  nB = _nB.data
+  v = _v.data
+
+  p = sortperm(nB; by=entropy)
+  ns[i] = nB[p[1]]
+  to_break = false
+  return ConservingNf(ns), to_break
+end
+
+function process_populations!(_ns::Pairing, _nB::Pairing, _v::Pairing, i::Int)
+  p = Int[]
+  ns = _ns.data
+  nB = _nB.data
+  v = _v.data
+
+  p = sortperm(nB)
+  n1 = nB[first(p)]
+  n2 = nB[last(p)]
+  to_break = false
+  ns[2 * i] = n1
+  ns[2 * i - 1] = n2
+  if length(v) == 2
+    to_break = true
```

Comment:
I think it would be better to make this `to_break` logic a separate function outside of `process_populations!`